### PR TITLE
Add some lr schedulers with warmup

### PR DESCRIPTION
**Pull request by LiuChiachi**
Add learning rate scheduler with warmup which increases or decreases initial learning rate linearly.
新增API对模型无影响的验证：
glue.py上使用两个lr scheduler（SST-2上），前100个steps可以对齐（少数误差>=1e-6)，3个epoch后验证集上的acc 误差+- 0.6%之内

### CODE REVIEW

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+            from paddlenlp.transformers import LinearSchedulerWithWarmup
+            lr, warmup_steps, max_steps = 0.1, 100, 1000
+            lr_scheduler = LinearSchedulerWithWarmup(lr, max_steps, warmup_steps)
+
+    """
+
+    def __init__(self,
+                 learning_rate,
+                 num_training_steps,
+                 num_warmup_steps,
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py]: 这里参数名简单些吧，可以用`warmup, total_steps`，另外`num_warmup_steps`换成`warmup`也希望同时支持设置integer和float以按照step数和比例进行warmup。其他几个scheduler相同。

Comment (guoshengCS) [reply PaddleNLP/paddlenlp/transformers/optimization.py]: 注意Python2的integer要包含int和long

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+                return float(current_step) / float(max(1, num_warmup_steps))
+            return max(0.0,
+                       float(num_training_steps - current_step) /
+                       float(max(1, num_training_steps - num_warmup_steps)))
+
+        super(LinearSchedulerWithWarmup, self).__init__(
+            learning_rate, lr_lambda, last_epoch, verbose)
+
+
+class ConstantSchedulerWithWarmup(LambdaDecay):
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py]: 类名考虑下能否简化或者加些别名，比如这个`ConstantSchedulerWithWarmup`里的constant可以用const

File: PaddleNLP/paddlenlp/transformers/optimization.py (hunk 1 above)
Comment (LiuChiachi) [reply PaddleNLP/paddlenlp/transformers/optimization.py]: 十分感谢~Done.

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+            progress = float(current_step - warmup_steps) / float(
+                max(1, total_steps - warmup_steps))
+            return max(0.0, 0.5 * (
+                1.0 + math.cos(math.pi * float(num_cycles) * 2.0 * progress)))
+
+        super(CosineSchedulerWithWarmup, self).__init__(
+            learning_rate, lr_lambda, last_epoch, verbose)
+
+
+class CosineWithHardRestartsScheduleWithWarmup(LambdaDecay):
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py]: 看这`CosineWithHardRestartsScheduleWithWarmup`名字实在太长，需要简化下，另外注意到这个和上面的`CosineSchedulerWithWarmup`相差不大，只是多了下面这一步
```python
            if progress >= 1.0:
                 return 0.0
```
可否将这两个合并通过增加参数来区别使用，一方面简化这个的名字，一方面减少接口数目。

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+            if progress >= 1.0:
+                return 0.0
+            return max(0.0, 0.5 * (1.0 + math.cos(math.pi * (
+                (float(num_cycles) * progress) % 1.0))))
+
+        super(CosineWithHardRestartsScheduleWithWarmup, self).__init__(
+            learning_rate, lr_lambda, last_epoch, verbose)
+
+
+class PolynomialDecaySchedulerWithWarmup(LambdaDecay):
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py]: 可以参考这里https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/schedulers.py

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+]
+
+
+def is_integer(number):
+    if sys.version > '3':
+        return isinstance(number, int)
+    return isinstance(number, (int, long))
+
+
+class LinearSchedulerWithWarmup(LambdaDecay):
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py]: 如果要保持`LinearSchedulerWithWarmup`这种命名方式的话，个人建议将这些class name中的Scheduler都改为Decay，一方面更简短，另一方面表意也更清晰是在decay的基础上加上了warmup，也和Paddle内置的大部分scheduler命名一致 https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/optimizer/lr.py#L21 ，
```python
__all__ = [
    'LRScheduler', 'NoamDecay', 'PiecewiseDecay', 'NaturalExpDecay',
    'InverseTimeDecay', 'PolynomialDecay', 'LinearWarmup', 'ExponentialDecay',
    'MultiStepDecay', 'StepDecay', 'LambdaDecay', 'ReduceOnPlateau',
    'CosineAnnealingDecay'
]
```
只是在这些decay的基础上多了warmup的过程

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+            lr_scheduler = ConstSchedulerWithWarmup(lr, warmup_steps)
+
+    """
+
+    def __init__(self, learning_rate, warmup, last_epoch=-1, verbose=False):
+        warmup_steps = warmup
+
+        def lr_lambda(current_step):
+            if current_step < warmup_steps:
+                return float(current_step) / float(max(1.0, warmup_steps))
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py]: 这里的warmup_steps是否也需要和其他一样需要换算

**Review by ZeyuChen — APPROVED** [review]
LGTM，@guosheng 最后决定合入，目前的API风格感觉我觉得不错。

File: PaddleNLP/examples/glue/run_glue.py [hunk 7]
```diff
@@ -355,17 +356,10 @@ def do_train(args):
...
-    lr_scheduler = paddle.optimizer.lr.LambdaDecay(
-        args.learning_rate,
-        lambda current_step, num_warmup_steps=warmup_steps,
-        num_training_steps=num_training_steps : float(
-            current_step) / float(max(1, num_warmup_steps))
-        if current_step < num_warmup_steps else max(
-            0.0,
-            float(num_training_steps - current_step) / float(
-                max(1, num_training_steps - num_warmup_steps))))
+    warmup_steps = args.warmup_steps if args.warmup_steps > 0 else args.warmup_proportion
```
Comment (guoshengCS) [inline PaddleNLP/examples/glue/run_glue.py]: 按照这里新增的这些scheduler，这里可以直接传入使用args.warmup_proportion

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+# limitations under the License.
+
+import sys
+import math
+
+from paddle.optimizer.lr import LambdaDecay
+
+__all__ = [
+    'LinearDecayWithWarmUp', 'ConstSchedulerWithWarmup',
+    'CosineDecayWithWarmUp', 'PolyDecayWithWarmUp'
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py]: WarmUp还是改用Warmup吧，这个还是以整体作为一个词使用居多，如这里文档中多次给出的那样

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import math
+
+from paddle.optimizer.lr import LambdaDecay
+
+__all__ = [
+    'LinearDecayWithWarmUp', 'ConstSchedulerWithWarmup',
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py]: `ConstSchedulerWithWarmup`用`ConstScheduleWithWarmup`吧

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+
+        super(LinearDecayWithWarmUp, self).__init__(learning_rate, lr_lambda,
+                                                    last_epoch, verbose)
+
+
+class ConstSchedulerWithWarmup(LambdaDecay):
+    """
+    Create a learning rate scheduler, which increases learning rate linearly
+    from 0 to given `learning_rate` during warmup periods and keeps learning
+    rate a constant.
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py]: keeps learning rate as a constant after that.

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+
+    Args:
+        learning_rate (float): The base learning rate. It is a python float
+            number.
+        warmup (int): The number of steps for warmup.
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py]: 同其他warmup的文档，这里应该漏掉了proportion的方式

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+        # Input check
+        if num_cycles is not None:
+            assert not with_hard_restarts and isinstance(num_cycles, int) or with_hard_restarts and isinstance(num_cycles, float), \
+            "`num_circles` should be an integer while `with_hard_restarts` is False, an float while `with_hard_restarts` is True."
+
+        def lr_lambda(current_step):
+            if current_step < warmup_steps:
+                return float(current_step) / float(max(1, warmup_steps))
+            if num_cycles is None:
+                num_cycles = 1 if not with_hard_restarts else 0.5
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py]: 放到lr_lambda外边

File: PaddleNLP/paddlenlp/transformers/optimization.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+    def __init__(self,
+                 learning_rate,
+                 total_steps,
+                 warmup,
+                 lr_end=1e-7,
+                 power=1.0,
+                 last_epoch=-1,
+                 verbose=False):
+        lr_init = learning_rate
+        assert lr_init > lr_end, f"`lr_end` must be be smaller than `learning_rate`. But `lr_end` is {lr_end} while `learning_rate` is {lr_init}."
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/transformers/optimization.py:243]: 注意```But `lr_end` is {lr_end} while `learning_rate` is {lr_init}.```这里的内容是否会有问题

File: PaddleNLP/examples/glue/run_glue.py (hunk 7 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/glue/run_glue.py]: `run_glue`同时有`warmup_steps`和`warmup_proportion`参数，`warmup_steps`的参数解释是`"Linear warmup over warmup_steps. If > 0: Override warmup_proportion"`。这里还是否需要直接传入`args.warmup_proportion`呢

File: PaddleNLP/paddlenlp/transformers/optimization.py (hunk 13 above)
Comment (LiuChiachi) [reply PaddleNLP/paddlenlp/transformers/optimization.py:243]: For example: image
这样是否可以呢

**Review by guoshengCS — APPROVED** [review]
